src/diffusion_for_multi_scale_molecular_dynamics/utils/d3pm_utils.py

```diff
@@ -0,0 +1,49 @@
+"""Common operations used for Discrete Diffusion."""
+import einops
+import torch
+
+
+def class_index_to_onehot(x: torch.Tensor, num_classes: int) -> torch.Tensor:
+    """Convert a tensor of class indices to a one-hot representation.
+
+    Args:
+        x: long tensor to encode
+        num_classes: total number of classes
+
+    Returns:
+        float tensor of 0s and 1s. The size is x.size() + (num_classes)
+    """
+    # the last .to() acts on the tensor type to avoid longs
+    return torch.nn.functional.one_hot(x.long(), num_classes=num_classes).to(x)
+
+
+def compute_q_xt_bar_xo(one_hot_x0: torch.Tensor, q_bar_t: torch.Tensor) -> torch.Tensor:
+    """Compute q(x_t | x_0).
+
+    This is done by the vector-matrix product: x_0 \bar{Q}_t assuming x_0 is a one-hot vector or a distribution over
+    different classes.
+
+    Args:
+        one_hot_x0: initial state (x_0). The last dimension should be the number of classes.
+        q_bar_t: cumulative Markov transition matrix (\bar{Q}_t). The last 2 dimensions should be the number of classes.
+
+    Returns:
+        matrix-vector product between one_hot_x0 and q_bar_t that defines q(x_t | x_0)
+    """
+    return einops.einsum(one_hot_x0.to(q_bar_t), q_bar_t, "... j, ... j i -> ... i")
+
+
+def compute_q_xt_bar_xtm1(one_hot_xt: torch.Tensor, q_t: torch.Tensor) -> torch.Tensor:
+    """Compute q(x_t | x_{t-1}).
+
+    This is done by the vector-matrix product: x_t Q_t^T assuming x_t is a one-hot vector or a distribution over
+    different classes.
+
+    Args:
+        one_hot_xt: state (x_t). The last dimension should be the number of classes.
+        q_t: Markov transition matrix (Q_t). The last 2 dimensions should be the number of classes.
+
+    Returns:
+        matrix-vector product between one_hot_xt and q_t^T that defines q(x_t | x_{t-1})
+    """
+    return einops.einsum(one_hot_xt.to(q_t), torch.transpose(q_t, -2, -1), "... j, ... i j -> ... i")
```

Comment:
The formula where this appears in the overleaf is Eq. 77. 
I think the formula should be

$$ \big[a_{t} \cdot Q_{t}^T \big] \_{i} = \sum_{j=1}^{K} \big[a_{t}\big]\_{j} \big[Q_{t}\big]\_{ij}$$

However, the formula implemented in the code right now is

$$ output\_{i} = \sum_{j=1}^{K} \big[a_{t}\big]\_{j} \big[Q_{t}\big]\_{ji}$$

I may have changed the overleaf since you wrote this. I may have introduced an error... We should align on what the correct formula should be here.

Reply:
No sure this has been addressed. Closing for now, revisit in the future.